rust_src/src/buffers.rs

```diff
@@ -1824,6 +1827,138 @@ pub fn rename_buffer(newname: LispStringRef, unique: LispObject) -> LispStringRe
     ThreadState::current_buffer_unchecked().name_.into()
 }
 
+/// Set the endpoints of OVERLAY to BEG and END in BUFFER.
+/// If BUFFER is omitted, leave OVERLAY in the same buffer it inhabits now.
+/// If BUFFER is omitted, and OVERLAY is in no buffer, put it in the current
+/// buffer.
+#[lisp_fn(min = "3")]
+pub fn move_overlay(
+    overlay: LispObject,
+    beg: LispObject,
+    end: LispObject,
+    buffer: LispObject,
+) -> LispObject {
+    let count = c_specpdl_index();
+    let mut overlay_ref = LispOverlayRef::from(overlay);
+    let mut buf = buffer
+        .as_buffer()
+        .or_else(|| overlay_buffer(overlay_ref))
+        .unwrap_or_else(ThreadState::current_buffer_unchecked);
+
+    if !buf.is_live() {
+        error!("Attempt to move overlay to a dead buffer");
+    }
+
+    if beg.is_marker() && !buf.eq(&marker_buffer(beg.into()).unwrap()) {
+        xsignal!(Qerror, "Marker points into wrong buffer", beg);
+    }
+
+    if end.is_marker() && !buf.eq(&marker_buffer(end.into()).unwrap()) {
+        xsignal!(Qerror, "Marker points into wrong buffer", end);
+    }
+
+    let beg_num = LispNumber::from(beg);
+    let end_num = LispNumber::from(end);
+    let (beg_num, end_num) = if beg_num.to_fixnum() > end_num.to_fixnum() {
+        (end_num, beg_num)
+    } else {
+        (beg_num, end_num)
+    };
+    unsafe { specbind(Qinhibit_quit, Qt) };
+    let obuffer = overlay_buffer(overlay_ref);
+    let mut o_beg = None;
+    let mut o_end = None;
+    if let Some(mut ob) = obuffer {
+        o_beg = overlay_start(overlay_ref);
+        o_end = overlay_end(overlay_ref);
+        unsafe { unchain_both(ob.as_mut(), overlay_ref.into()) };
+    }
```

Comment:
Alternatively, this could be written:

```
let (o_beg, o_end) = match obuffer {
    Some(mut ob) => {
        let o_beg = overlay_start(overlay_ref);
        let o_end = overlay_end(overlay_ref);
        unsafe { unchain_both(ob.as_mut(), overlay_ref.into()) };
        (o_beg, o_end)
    },
    None => (None, None)
};
```

Reply:
Oh, that's much better, thank you!